File: scripts/robotscan_client.js
const net = require('net');

/**
 * RobotScan TCP Client for OpenClaw (JavaScript version)
 * Handles "msgBegin{JSON}msgEnd" protocol.
 */

const TARGET_IP = "127.0.0.1";
const TARGET_PORT = 18878;

const action_map = {
  "start": "CommModule_Start",
  "stop": "CommModule_Stop",
  "suspend": "CommModule_Suspend",
  "resume": "CommModule_Resume",
  "open": "CommModule_Open_Template",
};

async function callRobot(action, params = {}) {
  const cmd_name = action_map[action] || action;
  const payload = {
    topicWorkflow: "ExtraCommModuleRequest",
    redirectAutomationCmd: "1",
    cmd: cmd_name,
    ...params
  };

  const message = `msgBegin${JSON.stringify(payload)}msgEnd`;

  const result = {
    success: false,
    action: action,
    sent_payload: payload,
    response: null,
    error: null
  };

  return new Promise((resolve) => {
    const client = new net.Socket();
    client.setTimeout(2000); // Only timeout if connection or sending fails

    client.connect(TARGET_PORT, TARGET_IP, () => {
      // Send the command and resolve once it's written to the buffer.
      // Status information is written to log files by RC, not via TCP response.
      client.write(message, () => {
        result.success = true;
        client.destroy();
      });
    });

    client.on('error', (err) => {
      result.error = err.message;
      resolve(result);
    });

    client.on('timeout', () => {
      result.error = "Connection/Send Timeout";
      client.destroy();
    });

    client.on('close', () => {
      resolve(result);
    });
  });
}

// CLI handle
if (require.main === module) {
  const args = process.argv.slice(2);
  let action = "";
  let params = {};

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--action") action = args[i + 1];
    if (args[i] === "--params") {
      try {
        params = JSON.parse(args[i + 1]);
      } catch (e) {
        process.stdout.write(JSON.stringify({ success: false, error: "Invalid JSON in --params" }));
        process.exit(1);
      }
    }
  }

  if (!action) {
    process.stdout.write(JSON.stringify({ success: false, error: "Missing --action" }));
    process.exit(1);
  }

  callRobot(action, params).then(res => {
    process.stdout.write(JSON.stringify(res, null, 2));
  });
}

File: scripts/robotscan_client.py
import sys
import socket
import json
import argparse

# RobotScan 协议默认对端（可能是服务也可能是 RC 本身）在本地 18878 端口
TARGET_IP = "127.0.0.1"
TARGET_PORT = 18878

def call_robot(cmd_name, params=None):
    """
    通过 TCP Socket 发送协议指令并返回 JSON 结果。
    """
    if params is None:
        params = {}

    payload = {
        "topicWorkflow": "ExtraCommModuleRequest",
        "redirectAutomationCmd": "1",
        "cmd": cmd_name
    }
    payload.update(params)

    message = f"msgBegin{json.dumps(payload)}msgEnd"
    
    result = {
        "success": False,
        "action": cmd_name,
        "sent_payload": payload,
        "response": None,
        "error": None
    }

    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(5)
            s.connect((TARGET_IP, TARGET_PORT))
            s.sendall(message.encode('utf-8'))
            
            try:
                data = s.recv(4096).decode('utf-8', errors='ignore')
                if data:
                    # 简单剥离 msgBegin/msgEnd (如果存在)
                    clean_data = data.replace("msgBegin", "").replace("msgEnd", "")
                    try:
                        result["response"] = json.loads(clean_data)
                    except json.JSONDecodeError:
                        result["response"] = clean_data
                result["success"] = True
            except socket.timeout:
                result["error"] = "Timeout: No response from robot"
                # 有些指令可能没有即时回复
                result["success"] = True 
                
    except Exception as e:
        result["error"] = str(e)
    
    print(json.dumps(result, indent=2, ensure_ascii=False))
    return result

def main():
    parser = argparse.ArgumentParser(description="RobotScan TCP 客户端控制脚本")
    parser.add_argument("--action", type=str, required=True, 
                        choices=["start", "stop", "suspend", "resume", "open", "scan", "status"],
                        help="执行的动作 (start, stop, suspend, resume, open, scan, status)")
    parser.add_argument("--params", type=str, default="{}", 
                        help="JSON 格式的参数字符串")

    args = parser.parse_args()

    # 指令名映射 (根据协议文档)
    action_map = {
        "start": "CommModule_Start",
        "stop": "CommModule_Stop",
        "suspend": "CommModule_Suspend",
        "resume": "CommModule_Resume",
        "open": "CommModule_Open_Template",
        "scan": "CommModule_Start_Scan",  # 补充可能的扫描指令
        "status": "CommModule_Get_Status"  # 补充可能的状态查询指令
    }

    target_cmd = action_map.get(args.action, args.action)
    
    try:
        params_dict = json.loads(args.params)
    except json.JSONDecodeError:
        print(json.dumps({"success": False, "error": "Invalid JSON in --params"}, indent=2))
        sys.exit(1)

    call_robot(target_cmd, params_dict)

if __name__ == "__main__":
    main()